### PR TITLE
Add UTs for C#11 file-scoped types feature

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CastConcreteTypeToInterface.CSharp11.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+﻿using System;
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CastConcreteTypeToInterface.CSharp11.cs:1]: I think we don't need tests for this. The rule subscribes only to cast and `AsIs` expressions and there are no scope-related checks.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeAbstract.CSharp11.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeAbstract.CSharp11.cs:1]: As far as I can tell, this rule does not check the type modifiers. I think it's enough to have one very simple noncompliant case and one compliant.

**Comment by costin-zaharia-sonarsource** [comment]
I would add a test for `NonDerivedPrivateClassesShouldBeSealed`. I wonder if it brings any value to modify the rule and suggest to make sealed also the non-derived file-scoped types. The magic happens in `IsPrivateButNotSealedType`

**Comment by costin-zaharia-sonarsource** [comment]
Could you please add tests also for: `TypeMemberVisibility` (https://rules.sonarsource.com/csharp/RSPEC-3059)

e.g.
```
file class Class1
{
    public void Method()
    {
```

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]

**Comment by mary-georgiou-sonarsource** [comment]
> I would add a test for `NonDerivedPrivateClassesShouldBeSealed`. I wonder if it brings any value to modify the rule and suggest to make sealed also the non-derived file-scoped types. The magic happens in `IsPrivateButNotSealedType`

@costin-zaharia-sonarsource  File-scoped types have to be always top-level (cannot be nested) and they cannot also have access modifiers (cannot be private).

If they should be sealed in case they are not derived in the file where they are declared it's probably another rule. We can discuss if it worths creating one in case.

**Comment by costin-zaharia-sonarsource** [comment]
> Based on the fact that file-scoped types are not visible outside the class and they cannot use accessibility modifiers do we need this? Am I missing something?

I would document the behavior since the rule description contains:
```
There’s no point in having a public member in a non-public type because objects that can’t access the type will never have the chance to access the member.
```

But indeed, I think the rule does not need to be changed.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CastConcreteTypeToInterfaceTest.cs [hunk 3]
```diff
@@ -28,5 +28,6 @@ public class CastConcreteTypeToInterfaceTest
         [TestMethod]
         public void CastConcreteTypeToInterface() =>
             new VerifierBuilder<CastConcreteTypeToInterface>().AddPaths("CastConcreteTypeToInterface.cs").Verify();
+
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/CastConcreteTypeToInterfaceTest.cs:31]: pedantic: can be removed

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**